coverage/hare_coverage.py

```diff
@@ -0,0 +1,319 @@
+
+
+class Strategy(ABC):
+
+    """Abstract base class for coverage tool."""
+
+    @abstractmethod
+    def cleanup(self) -> None:
+        """
+        Cleaning up old coverage reports.
+        :return: None
+        """
+        pass
+
+    @abstractmethod
+    def run_ut(self) -> None:
+        """
+        Execute unit testcases.
+        :return: None
+        """
+        pass
+
+    @abstractmethod
+    def run_coverage(self) -> None:
+        """
+        Running coverage tool over UTs.
+        :return: None
+        """
+        pass
+
+    @abstractmethod
+    def check_report(self) -> None:
+        """
+        Validate the report generated
+        :return: None
+        """
+        pass
+
+
+class PythonCoverage(Strategy):
+
+    # This is source to test directory map. It will be used run coverage.
+    # All the newly added source and test should be added here for coverage calculation.
+    source_test_map: Dict[str, str] = {
+        "hax/hax": "hax/test/",
+        "provisioning/miniprov/hare_mp": "provisioning/miniprov/test/"
+    }
+
+    def __init__(self, cov_type: str = "xml", report_dir: str = "hare_coverage_report",
+                 source_test_map: Optional[Dict[str, str]] = None):
+        """
+        Initialize coverage params.
+        :param cov_type: coverage type xml or html
+        :param source_test_map: source module to test module map
+        """
+        self.cov_type: str = "html" if cov_type.lower() == "html" else "xml"
+        if source_test_map is None:
+            self.src_tst_map: Optional[Dict[str, str]] = self.source_test_map
+        else:
+            self.src_tst_map: Optional[Dict[str, str]] = source_test_map
+        self.report_dir = report_dir
+        self.report = self.report_dir + sep
+        self.report += "html_report" if self.cov_type == "html" else "coverage.xml"
+        self.utils = Utility()
+        logging.info("Running Python Coverage Tool.")
+
+    def cleanup(self) -> None:
+        """
+        Cleaning up old coverage reports.
+        :return: None
+        """
+        logging.info("Cleaning up the report.")
+        try:
+            if self.utils.is_exist(self.report_dir) is True:
+                self.utils.run_cmd(f"rm -rf {self.report_dir}")
+        except Exception as e:
+            logging.error(f"Error {e} while cleanup.")
+
+    def run_ut(self) -> None:
+        """
+        Execute unit testcases.
+        :return: None
+        """
+        logging.info("Running UT's.")
+        dir_scanner = DirectoryScanner("test")
+        for test_dir_path in dir_scanner.search_dir(getcwd()):
+            self.utils.run_cmd(f"pytest {test_dir_path}")
+
+    def run_coverage(self) -> None:
+        """
+        Running coverage tool over UTs.
+        :return: None
+        """
+        logging.info("Running Coverage.")
+        command_line: str
+        for index, (src, tst) in enumerate(self.src_tst_map.items()):

```

Comment:
mypy says: Item "None" of "Optional[Dict[str, str]]" has no attribute "items"

Reply:
Resolved